ipq40xx: add support for Linksys WHW01 v1

darkarnium opened the pull request:
This patch adds support for Linksys WHW01 v1 ("Velop") [FCC ID Q87-03331].
```
Specification
-------------

SOC:             Qualcomm IPQ4018
WiFi 1:          Qualcomm QCA4019 IEEE 802.11b/g/n
WiFi 2:          Qualcomm QCA4019 IEEE 802.11a/n/ac
Bluetooth:       Qualcomm CSR8811 (A12U)
Ethernet:        Qualcomm Atheros QCA8072 (2-port)
SPI Flash 1:     Mactronix MX25L1605D (2MB)
SPI Flash 2:     Winbond W25M02GV (256MB)
DRAM:            Nanya NT5CC128M16IP-DI (256MB)
LED Controller:  NXP PCA963x (I2C)
Buttons:         Single reset button (GPIO).

Notes
-----

There does not appear to be a way to trigger TFTP recovery without entering
U-Boot. The device must be opened to access the serial console in order to
first flash OpenWrt onto a device from factory.

The device has automatic recovery backed by a second set of partitions on
the larger of the two SPI flash ICs. Both the primary and secondary must
be flashed to prevent accidental rollback to "factory" after 3 failed boot
attempts.

Serial console
--------------

A serial console is available on the following pins of the populated J2
connector on the device mainboard (115200 8n1).

(<-- Top of PCB / Device)

  J2
  [o o o o o o]
       |   | |
       |   |  `-- GND
       |    `---- TX
       `--------- RX

Installation instructions
-------------------------

1. Setup TFTP server with server IP set to 192.168.1.236.
2. Copy compiled `...squashfs-factory.bin` to `nodes-jr.img` in tftp root.
3. Connect to console using pinout detailed in the serial console section.
4. Power on device and press enter when prompted to drop into U-Boot.
5. Flash first partition device via `run flashimg`.
6. Once complete, reset device and allow to power up completely.
7. Once comfortable with device upgrade reboot and drop back into U-Boot.
8. Flash the second partition (recovery) via `run flashimg2`.

Revert to "factory"
-------------------

1. Download latest firmware update from vendor support site.
2. Copy extracted `.img` file to `nodes-jr.img` in tftp root.
3. Connect to console using pinout detailed in the serial console section.
4. Power on device and press enter when prompted to drop into U-Boot.
5. Flash first partition device via `run flashimg`.
6. Once complete, reset device and allow to power up completely.
7. Once comfortable with device upgrade reboot and drop back into U-Boot.
8. Flash the second partition (recovery) via `run flashimg2`.
```
Signed-off-by: Peter Adkins <peter@sunkenlab.com>

adschm commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-whw01v1.dts:
```diff
@@ -0,0 +1,238 @@
+		};
+
+		unused {
+			label = "unused";
```
no color?

darkarnium replied:
I've added this in.

I'm not certain whether it is required as the device has a single RGB LED which is connected to the i2c controller; I'll need to read more into the use of the LED common bindings to better understand how the `color` property is used :)

adschm commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-whw01v1.dts:
```diff
@@ -0,0 +1,238 @@
+		};
+	};
+};
+
```
remove empty line at EOF.

adschm commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-whw01v1.dts:
```diff
@@ -0,0 +1,238 @@
+
+/ {
+	model = "Linksys WHW01v1";
+	compatible = "linksys,whw01v1";
```
If v1 stands for "version 1", please separate it by a space for the model and a hyphen for the compatible.

This will require corresponding adjustments to DTS file name and device definition etc.

darkarnium commented:
Thanks for the quick review, and the feedback! I've just added the requested changes and I am running a build now. Once complete I'll flash a test device to ensure no breaking changes and update this PR.

darkarnium commented:
Rebuild complete and new `...factory.bin` image flashed via TFTP without error (`openwrt-ipq40xx-generic-linksys_whw01-v1-squashfs-factory.bin`). Confirmed LEDs, ethernet switch, and radios are still functioning. 👍

darkarnium commented:
Rebuild complete and new `...factory.bin` image flashed via TFTP without error (`openwrt-ipq40xx-generic-linksys_whw01-v1-squashfs-factory.bin`). Pushed a change to remove the padding to `IMAGESIZE` in favour of just padding to `BLOCKSIZE` - which is all that is required. Also fixed up `ro` flags on a few partitions which shouldn't have been there.

Confirmed LEDs, ethernet switch, and radios are still functioning. Appologies for the change after sign-off, but I thought this would be better than an immediate patch being required. 👍

adschm commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-whw01-v1.dts:
```diff
@@ -0,0 +1,237 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+/dts-v1/;
```
dts-v1 can be dropped, it's already in qcom-ipq4018.dtsi. (This does not require any testing.)

darkarnium replied:
Removed.

adschm commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-whw01-v1.dts:
```diff
@@ -0,0 +1,237 @@
+	};
+
+	soc {
+		gpio_keys {
```
Just "keys" in root node?

darkarnium replied:
Moved to `keys`.

adschm commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-whw01-v1.dts:
```diff
@@ -0,0 +1,237 @@
+
+	soc {
+		gpio_keys {
+			pinctrl-0 = <0x52>;
```
What's that?

darkarnium replied:
Removed, I'm not sure how this got here, nice catch.

adschm commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-whw01-v1.dts:
```diff
@@ -0,0 +1,237 @@
+			pinctrl-names = "default";
+			compatible = "gpio-keys";
+
+			button@1 {
```
"reset"?

darkarnium replied:
Renamed to `reset` to match other devices.

adschm commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-whw01-v1.dts:
```diff
@@ -0,0 +1,237 @@
+
+			button@1 {
+				label = "reset";
+				linux,code = <0x198>;
```
Please use macros.

darkarnium replied:
Macros re-added, as below this was an artifact of sourcing some of this device-tree from the factory firmware run through `dtc`.

adschm commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-whw01-v1.dts:
```diff
@@ -0,0 +1,237 @@
+			button@1 {
+				label = "reset";
+				linux,code = <0x198>;
+				gpios = <&tlmm 0x3f 0x01>;
```
Please use numeric value for second entry and macro for GPIO_ACTIVE_LOW.

darkarnium replied:
Added.

adschm commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-whw01-v1.dts:
```diff
@@ -0,0 +1,237 @@
+				label = "reset";
+				linux,code = <0x198>;
+				gpios = <&tlmm 0x3f 0x01>;
+				linux,input-type = <0x01>;
```
Macro ...

darkarnium replied:
Resolved.

adschm commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-whw01-v1.dts:
```diff
@@ -0,0 +1,237 @@
+	cs-gpios = <&tlmm 54 GPIO_ACTIVE_HIGH>, <&tlmm 4 GPIO_ACTIVE_HIGH>;
+
+	nor@0 {
+		spi-max-frequency = <0x16e3600>;
```
really?

darkarnium replied:
Appologies, these nodes were originally taken from the device-tree sliced out of the factory update image and run through `dtc` as a result a lot of the raw values are in the wrong base or without macros. I'll fix these up now.

adschm commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-whw01-v1.dts:
```diff
@@ -0,0 +1,237 @@
+		partitions {
+			compatible = "fixed-partitions";
+			#address-cells = <0x01>;
+			#size-cells = <0x01>;
```
`<1>`

darkarnium replied:
Updated

adschm commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-whw01-v1.dts:
```diff
@@ -0,0 +1,237 @@
+
+			partition@0 {
+				label = "0:SBL1";
+				reg = <0x00 0x40000>;
```
0x0 instead of 0x00 ....

darkarnium replied:
Updated.

adschm commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-whw01-v1.dts:
```diff
@@ -0,0 +1,237 @@
+				read-only;
+			};
+
+			partition@e0000 {
```
d0000

darkarnium replied:
These `partition` blocks have all just been re-generated from the factory firmware. These should now be correct.

adschm commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-whw01-v1.dts:
```diff
@@ -0,0 +1,237 @@
+				read-only;
+			};
+
+			partition@160000 {
```
170000

darkarnium replied:
These `partition` blocks have all just been re-generated from the factory firmware. These should now be correct.

adschm commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-whw01-v1.dts:
```diff
@@ -0,0 +1,237 @@
+				read-only;
+			};
+
+			partition@170000 {
```
And again? Maybe you should check again which partitioning is actually the right one ...

darkarnium replied:
These `partition` blocks have all just been re-generated from the factory firmware. These should now be correct.

adschm commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-whw01-v1.dts:
```diff
@@ -0,0 +1,237 @@
+			partition@1b0000 {
+				label = "devinfo";
+				reg = <0x1e0000 0x20000>;
+			};
```
These should probably be read-only as well?

darkarnium replied:
I'm unsure about this one, as it seems to contain some secrets which look generated. That said, I've marked as `RO` to be safe.

adschm commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-whw01-v1.dts:
```diff
@@ -0,0 +1,237 @@
+
+	nand@1 {
+		reg = <1>;
+		status = "ok";
```
There is no "ok", just "okay". Is this disabled somewhere? If not, just remove the status, okay is default.

darkarnium replied:
I've removed the `status` property entirely. As for `ok` versus `okay` that's good to know! I was unsure this one as there are a few `ok` present in `qcom-ipq4019-ap.dk01.1.dtsi` which I was cribbing from.

adschm replied:
I have a patch pending for that, but seems like qcom doesn't care. For the future, just look at the documentation.

darkarnium replied:
Hah, good to know. I'll do so in future :)

adschm commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-whw01-v1.dts:
```diff
@@ -0,0 +1,237 @@
+		reg = <1>;
+		status = "ok";
+		compatible = "spi-nand";
+		spi-max-frequency = <0x16e3600>;
```
See above.

darkarnium replied:
Converted to decimal.

adschm commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-whw01-v1.dts:
```diff
@@ -0,0 +1,237 @@
+		partitions {
+			compatible = "fixed-partitions";
+			#address-cells = <0x01>;
+			#size-cells = <0x01>;
```
See above.

darkarnium replied:
Converted to decimal.

adschm commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-whw01-v1.dts:
```diff
@@ -0,0 +1,237 @@
+
+			partition@0 {
+				label = "kernel";
+				reg = <0x00 0x5000000>;
```
See above.

darkarnium replied:
These `partition` blocks have all just been re-generated from the factory firmware. These should now be correct. `0x0` has been zero padded to the same length as the others to make it easier to follow - per your other comment.

adschm commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-whw01-v1.dts:
```diff
@@ -0,0 +1,237 @@
+
+			partition@600000 {
+				label = "rootfs";
+				reg = <0x600000 0x4a00000>;
```
Actually, it would be better to use 7 digits for all values inside reg here to improve readability (for the nand).

darkarnium replied:
These `partition` blocks have all just been re-generated from the factory firmware. These should now be correct. Padded to the same number of digits.

adschm commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-whw01-v1.dts:
```diff
@@ -0,0 +1,237 @@
+		compatible = "nxp,pca9633";
+		#address-cells = <0x01>;
+		#size-cells = <0x00>;
+		reg = <0x62>;
```
See above.

darkarnium replied:
Resolved. `reg` has been left as hexadecimal as this is the default address of of the PCA9633 on the I2C bus - and is referenced in `drivers/legs/leds-pca963x.c` in hex.

adschm commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-whw01-v1.dts:
```diff
@@ -0,0 +1,237 @@
+
+		system_red {
+			label = "red:system";
+			color = <LED_COLOR_ID_RED>;
```
I'm not sure whether using color without function and in parallel to label makes any sense. I'd consider dropping the color property here.

darkarnium replied:
Dropped.

adschm commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-whw01-v1.dts:
```diff
@@ -0,0 +1,237 @@
+		unused {
+			label = "unused";
+			reg = <0x03>;
+			linux,default-trigger = "none";
```
This line can probably be dropped.

darkarnium replied:
Removed.

adschm commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-whw01-v1.dts:
```diff
@@ -0,0 +1,237 @@
+		};
+
+		unused {
+			label = "unused";
```
Missing color.

darkarnium replied:
Removed `unused` block entirely, as it isn't physically connected to any pins on the RGB LED, nor any other device.

adschm commented on target/linux/ipq40xx/image/Makefile:
```diff
@@ -557,6 +557,24 @@ define Device/linksys_mr8300
+	DEVICE_VENDOR := Linksys
+	DEVICE_MODEL := WHW01
+	DEVICE_VARIANT := v1
+	BOARD_NAME := whw01-v1
```
This can probably be dropped.

darkarnium replied:
Removed.

darkarnium replied:
Removed.

adschm commented on target/linux/ipq40xx/image/Makefile:
```diff
@@ -557,6 +557,24 @@ define Device/linksys_mr8300
+	PAGESIZE := 2048
+	UBINIZE_OPTS := -E 5    # EOD marks to "hide" factory sig at EOF
+	IMAGES += factory.bin
+	IMAGE/factory.bin  := append-kernel | pad-to $$$$(KERNEL_SIZE) | append-ubi | linksys-image type=WHW01 | pad-to $$$$(PAGESIZE) | check-size
```
Please properly wrap after max. 74 characters.

darkarnium replied:
Done.

adschm commented on target/linux/ipq40xx/image/Makefile:
```diff
@@ -557,6 +557,24 @@ define Device/linksys_mr8300
+	UBINIZE_OPTS := -E 5    # EOD marks to "hide" factory sig at EOF
+	IMAGES += factory.bin
+	IMAGE/factory.bin  := append-kernel | pad-to $$$$(KERNEL_SIZE) | append-ubi | linksys-image type=WHW01 | pad-to $$$$(PAGESIZE) | check-size
+	DEVICE_PACKAGES := uboot-envtools kmod-leds-pca963x kmod-bluetooth
```
I don't think kmod-bluetooth is typically installed by default. Can you provide any examples?

darkarnium replied:
Only one: `Device/meraki_mr32` in `bcm53xx`. Happy to remove if prefered.

darkarnium replied:
I've removed this kmod as the end-user can always re-add if desired.

darkarnium commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-whw01-v1.dts:
```diff
@@ -0,0 +1,237 @@
+			partition@a200000 {
+				label = "syscfg";
+				reg = <0xa200000 0x5e00000>;
+			};
```
`sysdiag` appears to be empty on all three of my test devices (filled with 0xFF), the usage unknown. I'll re-add for `sysdiag`.

`syscfg` is a UBIFS and contains a number of backup files and SQLite databases used by the factory firmware - such as databases of all MAC addresses ever seen by the device, etc. This is recreated if I flash back the factory firmware, so allowing RW allows the end-user to replace / re-use as desired - or at least remove the SQLite databases which contain device information about clients which persist between upgrades.

adschm replied:
Well, if it's factory stuff I would say it should not be altered by the OpenWrt user. However, you saying "it's recreated" means whatever you do here with OpenWrt, it will be wiped when installing OEM firmware anyway?

darkarnium replied:
That's correct. The factory firmware recreates the contents of this UBIFS. If it **does** exist, it continues to use the database files on the volume between firmware versions. These SQLite databases are used for what could only be described as "long lived telemetry" about devices seen by the units.

darkarnium commented:
Thanks again for the review and the feedback! I've just pushed a new commit, but not squashed / rebased so that the GitHub PR so that we don't lose track of the conversation history.

If you're happy with the changes / comments, please let me know and I can do a rebase and force push to squash :)

robimarko commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-whw01-v1.dts:
```diff
@@ -0,0 +1,237 @@
+		bootargs-append = " root=/dev/ubiblock0_0";
+	};
+
+	reserved-memory {
```
Are these needed?

darkarnium replied:
I'm unsure how to accurately confirm whether this is the case; I do not have access to QCOM datasheets for this board or SoC.

In the interest of caution these sections have been carved out based on the same `reserved-memory` sections present factory firmware device-tree. That said, these may have been for a Kernel built from a QCOM tree / BSP, and potentially for drivers not present in mainline? That said, it appears that the these additional sections were attempted to be upstreamed into `qcom-ipq4019-ap.dk01.1.dtsi` [previously](https://patchwork.kernel.org/project/linux-arm-kernel/patch/1521193101-4586-5-git-send-email-sricharan@codeaurora.org/).

An upper portion of `rsvd2` is already present in `qcom-ipq4019.dtsi` (`0x87e00000` to `0x88000000`), so this is only allocated as `0x300000` rather than `0x500000` - as the existing `smem` and `tz` entries in `qcom-ipq4019.dtsi` comprise the rest.

robimarko replied:
You can simply remove them and try using the device, if it does not crash during that then these are not needed.
Usually, some of these are left for the crash log which the mainline does not use while others are already upstream.

darkarnium replied:
Removing these entries and rebuilding the image doesn't appear to have broken any functionality, so I guess they're not needed. I've pushed a new commit with them removed.

pepe2k commented:
@darkarnium 

```
There does not appear to be a way to trigger TFTP recovery without entering
U-Boot. The device must be opened to access the serial console in order to
first flash OpenWRT onto a device from factory.
```
Please fix the project name in your commit description, see:
https://forum.openwrt.org/t/what-does-wrt-stand-for/27074/5?u=pepe2k

Cheers,
Piotr

darkarnium commented:
Sorry about that! I've just pushed a commit to resolve, updated the PR description, and also updated the wiki article in the OpenWrt ToH pages - as they had bits copied and pasted from my commit message.

Thanks and nice spot :)

darkarnium commented:
I've just pushed a commit to improve support for this device. the 5GHz radio had some issues after testing for a few days, which was resolved by extracting the BDFs from the vendor firmware. These have also been sent upstream.

Additionally, the upgrade process has been instructed to use the "Linksys" upgrade scripts, and the `kernel` and `alt_kernel` were erroneously marked as read-only. This has been resolved in order to allow sysupgrade to complete without error.

The device is working well in an 802.11s mesh without any significant impediments that I've seen since these changes.

predators46 commented:
is there any progress? i want to try

darkarnium commented:
Hello there,

I've just merged the current `master` branch back into this branch as there were a few conflicting changes I've since resolved. I was wondering if there's any changes left required from my end as part of this PR?

Thank you again for your time and assistance!

Regards,
Peter

adschm commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-whw01-v1.dts:
```diff
@@ -0,0 +1,211 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+
+#include "qcom-ipq4019-ap.dk01.1.dtsi"
```
I don't think using a device-specific DTSI here, and even across 4019 vs. 4018 platforms, is appropriate.

darkarnium replied:
The use of `qcom-ipq4019-ap.dk01.1.dtsi` as a base was done in-line with the patches present in the vendor's GPL package. That said, I've moved this to use `qcom-ipq4019.dtsi` now - which required quite a few additional nodes be added.

adschm commented:
Ping.

adschm commented:
... and the two patches need to be squashed.

darkarnium commented:
I'll take a look at this in a couple of days.

darkarnium commented:
I've pulled in the latest changes from `master` and resolved any conflicts. I've also recompiled from this branch and flashed both a `sysupgrade` and `factory` image onto separate WHW01 v1 to ensure that both images are working without issue.

adschm commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-whw01-v1.dts:
```diff
@@ -0,0 +1,323 @@
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *
+ */
```
This should be replace by a proper SPDX license identifier.

adschm commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-whw01-v1.dts:
```diff
@@ -0,0 +1,323 @@
+&wifi1 {
+	status = "okay";
+	qcom,ath10k-calibration-variant = "linksys-whw01-v1";
+};
```
missing newline at EOF.

adschm commented on target/linux/ipq40xx/image/generic.mk:
```diff
@@ -584,6 +584,26 @@ define Device/linksys_mr8300
+	PAGESIZE := 2048
+	UBINIZE_OPTS := -E 5    # EOD marks to "hide" factory sig at EOF
+	IMAGES += factory.bin
+	IMAGE/factory.bin  := append-kernel | pad-to $$$$(KERNEL_SIZE) | \
```
remove one space before :=

adschm commented:
I don't think anybody will properly review this as long as it's marked WIP ...

darkarnium commented:
Are there any other issues with this PR that need to be resolved? I've had several WHW01v1 devices deployed using this branch for the last 9 months or so without issue.

xzyk210957 commented:
Hi, I see now it's marked as 'PR pending' on openwrt table of hardware. I have a couple of these units which have been half-bricked due to person i bought them from having tried unverified versions of openwrt. U-boot is still fine but it's obvious from the 'printenv' that it's loading openwrt firmware and overwriting with oem firmware via 'run flashimg' is not booting and hangs on 'Starting Kernel'. I would like to try this firmware to see if it can solve the problem, otherwise it's going to take diving into u-boot issues. I can't find anywhere to download 'openwrt-ipq40xx-generic-linksys_whw01-v1-squashfs-factory.bin' from so I guess will have to wait until the PR is approved and it goes into master.

xzyk210957 commented:
Tried to compile from github.com/darkarnium/openwrt.git and got the following errors:
# Below error when executing 'make download and 'make''
> make download
make[2]: Leaving directory '/home/andy/openwrt/openwrt/feeds/packages/libs/dtc'
time: package/feeds/packages/dtc/download#0.17#0.06#0.20
    ERROR: package/feeds/packages/dtc failed to build.
make[1]: *** [package/Makefile:116: package/feeds/packages/dtc/download] Error 1

> make -j $(($(nproc)+1))
make[3] -C package/libs/gettext-full host-compile
make[3] -C package/libs/libiconv host-compile
make[3] -C feeds/packages/libs/libical compile
make[3] -C feeds/packages/libs/dtc compile
make[3] -C feeds/packages/utils/hwdata compile
    ERROR: package/feeds/packages/dtc failed to build.
make -r world: build failed. Please re-run make with -j1 V=s or V=sc for a higher verbosity level to see what's going on
make: *** [/home/andy/openwrt/openwrt/include/toplevel.mk:230: world] Error 1

#File ...../include/toplevel.mk at line 230 shown below
cp .config tmp/.config; \
Let me know if you need additional info..

PolynomialDivision commented:
Please also add the DSA conversion:
https://github.com/openwrt/openwrt/pull/4721

sniff122 commented:
is there any update on this pr? would love to be able to ditch the stock firmware on my 4 APs

xzyk210957 commented:
Hi, i ditched this in the end cause i tried so many things i probably ended up corrupting most of my partitions. Never did manage to get it to load, only ever got as far as 'loading kernel' and then it would hang.
Thinking about it later what's the point of buying the velop ones if you're going to put openwrt mesh system on it? OpenWrt has support for mesh for a while so why not but a cheaper (than linksys velop) router, or APs, and use those in an openwrt mesh, it will be a little cheaper.

11 May 2022 21:28:16 Lewis L. Foster ***@***.***>:

> is there any update on this pr? would love to be able to ditch the stock firmware on my 4 APs
> 
> —
> Reply to this email directly, view it on GitHub[https://github.com/openwrt/openwrt/pull/3682#issuecomment-1124265212], or unsubscribe[https://github.com/notifications/unsubscribe-auth/ASEB2LW2SWUOTL5SQJT3VJ3VJQJ57ANCNFSM4UZQISGA].
> You are receiving this because you commented.[Tracking image][https://github.com/notifications/beacon/ASEB2LRS2ELF4SLPUJ6QBETVJQJ57A5CNFSM4UZQISGKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOIMBOZ7A.gif]
>

sniff122 commented:
Was able to get it to boot on one of my nodes but I couldn't get LuCI into the image, I probably just haven't read the docs properly or something or I am forgetting a dependency (something about luasrcdiet)

xzyk210957 commented:
Good luck to you..i also got 4 of these running and to be honest they've been reasonably solid over the past 18 months, occasional loss of connection between master and child nodes but usually they reconnect.
I've decided to keep to use them Linksys firmware although they are in bridged mode so not my main router, fort that i have a linksys WRT3200 running openwrt and it's been rock solid.

11 May 2022 23:48:04 Lewis L. Foster ***@***.***>:

> Was able to get it to boot on one of my nodes but I couldn't get LuCI into the image, I probably just haven't read the docs properly or something or I am forgetting a dependency (something about luasrcdiet)
> 
> —
> Reply to this email directly, view it on GitHub[https://github.com/openwrt/openwrt/pull/3682#issuecomment-1124361386], or unsubscribe[https://github.com/notifications/unsubscribe-auth/ASEB2LWWSJCA5AOTOMIER5TVJQ2KJANCNFSM4UZQISGA].
> You are receiving this because you commented.[Tracking image][https://github.com/notifications/beacon/ASEB2LRDNV4HA67HFNRD4O3VJQ2KJA5CNFSM4UZQISGKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOIMCGJKQ.gif]
>

sniff122 commented:
ah, mine have been pretty stable on the stock firmware but ive wanted a bit more control, my current setup has 2 nodes wired and 2 on mesh and using pfsense in a proxmox VE (basically qemu/kvm on debian). Ive also got openwrt on a dlink dir-878 mainly just to play around and its great. Would definitely love this to become merged in though and i would try to help but im in over my head lmao

darkarnium commented:
Hey there,

I've had several of these APs running these changes for over a year now without any issues. The latest changes from this pull-request, before things floated out of date, are the ones I have in use. I'm a little surprised to hear of significant issues when flashing the devices, as although the initial load is a bit troublesome, per the "Installation instructions", once loaded upgrades have been okay. In terms of LuCI, I can also confirm that this is working without issue on the devices I have deployed.

In regards to the pull-request being stale: There have been several large changes to `master` since this pull-request was opened and the first batch of changes were submitted for review, and once again after adding in the requested changes. The changes from `master` were pulled back into this pull-request a couple of times, but it's since floated back out once again.

I'll need to look into the request to add the "DSA conversion" but I must admit that I'm a little hesitant to this and also pull back in the changes, if the pull-request is likely to go stale again. _That said_, I'll give it "one last hurrah" over the next few days and bring things in-line with `master`.

I should also note that I use 802.11r (fast-roaming), rather than 802.11s, in this configuration.


- Peter

sniff122 commented:
yeah i could probably do with looking more into compiling everything seeing if i can get it to compile with LuCI. Glad to hear its all been working for you, does give me some hope and its just me haha

xzyk210957 commented:
Hi Pete, you and i did speak about this when i was having problems and you offered to help, thanks for that.. but like i said i stuck with the official firmware since it's not too bad and operating in bridged mode makes it less important to have fuller control like openwrt gives me.

12 May 2022 00:03:25 Peter ***@***.***>:

> Hey there,
> 
> I've had several of these APs running these changes for over a year now without any issues. The latest changes from this pull-request, before things floated out of date, are the ones I have in use. I'm a little surprised to hear of significant issues when flashing the devices, as although the initial load is a bit troublesome, per the "Installation instructions", once loaded upgrades have been okay. In terms of LuCI, I can also confirm that this is working without issue on the devices I have deployed.
> 
> In regards to the pull-request being stale: There have been several large changes to *master* since this pull-request was opened and the first batch of changes were submitted for review, and once again after adding in the requested changes. The changes from *master* were pulled back into this pull-request a couple of times, but it's since floated back out once again.
> 
> I'll need to look into the request to add the "DSA conversion" but I must admit that I'm a little hesitant to this and also pull back in the changes, if the pull-request is likely to go stale again. /That said/, I'll give it "one last hurrah" over the next few days and bring things in-line with *master*.
> 
> I should also note that I use 802.11r (fast-roaming), rather than 802.11s, in this configuration.
> 
* > Peter
> 
> —
> Reply to this email directly, view it on GitHub[https://github.com/openwrt/openwrt/pull/3682#issuecomment-1124368557], or unsubscribe[https://github.com/notifications/unsubscribe-auth/ASEB2LVKROJGW4XWKKBECBDVJQ4DZANCNFSM4UZQISGA].
> You are receiving this because you commented.[Tracking image][https://github.com/notifications/beacon/ASEB2LV6JCINATMIWGGKL73VJQ4DZA5CNFSM4UZQISGKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOIMCIBLI.gif]
>

darkarnium commented:
@sniff122 once back in good shape, I'd be happy to send over my `.config` which should serve as a responable starting point, as I build with LuCI and a few other "bells and whistles".

@xzyk210957 Glad to hear that you found a pathway forward for your use case, even if it wasn't with OpenWrt! :)

sniff122 commented:
> @sniff122 once back in good shape, I'd be happy to send over my `.config` which should serve as a responable starting point, as I build with LuCI and a few other "bells and whistles".

that would be amazing

sniff122 commented:
I have a feeling something is up between the old version and the new feeds or something. I first had to get `include/meson.mk` and `include/openssl-engine.mk` from the master branch otherwise updating the feeds would spit out a load of errors about make files failing and looking at the logs it was saying about those files not existing.
Then when running `./scripts/feeds install -a` a load of warning messages is logged saying about no feed for package X (see file below).
[openwrt-build.txt](https://github.com/openwrt/openwrt/files/8674160/openwrt-build.txt)

Building works and it boots but basically no packages can be installed during the build process.

Tried on both my laptop running ubuntu 21.10 and a fresh proxmox ve container (LXC) running ubuntu 18.04.1 with the exact same results above

sniff122 commented:
Ok so the packages issue does appear to be because of the older build environment from the branch. I was able to manually modify the files to add the support for the WHW01 from a fresh clone of master and no more package issues. However when compiling, right at the end when its handling the device trees, the build fails with the following error
```
/home/lewis/openwrt/build_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/linux-ipq40xx_generic/linux-5.10.113/arch/arm/boot/dts/qcom-ipq4018-whw01-v1.dts:62.7-83.4: Warning (i2c_bus_reg): /soc/i2c@78b7000/leds: I2C bus unit address format error, expected "62"
```

sniff122 commented:
nevermind, turns out i had forgot to patch `target/linux/ipq40xx/patches-5.10/901-arm-boot-add-dts-files.patch`, looks like it built correctly, will test when i get home

sniff122 commented:
Forgot to post my update, flashed successfully and booted without any problems. Doesnt appear to have any major issues so thats a plus

darkarnium committed:
ipq40xx: add support for Linksys WHW01 v1

This patch adds support for Linksys WHW01 v1 ("Velop") [FCC ID Q87-03331].

Specification
-------------

SOC:             Qualcomm IPQ4018
WiFi 1:          Qualcomm QCA4019 IEEE 802.11b/g/n
WiFi 2:          Qualcomm QCA4019 IEEE 802.11a/n/ac
Bluetooth:       Qualcomm CSR8811 (A12U)
Ethernet:        Qualcomm QCA8072 (2-port)
SPI Flash 1:     Mactronix MX25L1605D (2MB)
SPI Flash 2:     Winbond W25M02GV (256MB)
DRAM:            Nanya NT5CC128M16IP-DI (256MB)
LED Controller:  NXP PCA963x (I2C)
Buttons:         Single reset button (GPIO).

Notes
-----

There does not appear to be a way to trigger TFTP recovery without entering
U-Boot. The device must be opened to access the serial console in order to
first flash OpenWrt onto a device from factory.

The device has automatic recovery backed by a second set of partitions on
the larger of the two SPI flash ICs. Both the primary and secondary must
be flashed to prevent accidental rollback to "factory" after 3 failed boot
attempts.

Serial console
--------------

A serial console is available on the following pins of the populated J2
connector on the device mainboard (115200 8n1).

(<-- Top of PCB / Device)

  J2
  [o o o o o o]
       |   | |
       |   |  `-- GND
       |    `---- TX
       `--------- RX

Installation instructions
-------------------------

1. Setup TFTP server with server IP set to 192.168.1.236.
2. Copy compiled `...squashfs-factory.bin` to `nodes-jr.img` in tftp root.
3. Connect to console using pinout detailed in the serial console section.
4. Power on device and press enter when prompted to drop into U-Boot.
5. Flash first partition device via `run flashimg`.
6. Once complete, reset device and allow to power up completely.
7. Once comfortable with device upgrade reboot and drop back into U-Boot.
8. Flash the second partition (recovery) via `run flashimg2`.

Revert to "factory"
-------------------

1. Download latest firmware update from vendor support site.
2. Copy extracted `.img` file to `nodes-jr.img` in tftp root.
3. Connect to console using pinout detailed in the serial console section.
4. Power on device and press enter when prompted to drop into U-Boot.
5. Flash first partition device via `run flashimg`.
6. Once complete, reset device and allow to power up completely.
7. Once comfortable with device upgrade reboot and drop back into U-Boot.
8. Flash the second partition (recovery) via `run flashimg2`.

Signed-off-by: Peter Adkins <peter@sunkenlab.com>

darkarnium committed:
Merge branch 'master' of https://github.com/openwrt/openwrt into Linksys-WHW01v1

darkarnium commented:
Hey @sniff122,

Thanks for the update! I just reapplied this change set against the current `master` and ran a clean build. The only manual change to the config (`make menuconfig`) was the inclusion of LuCI.

The builds have complete cleanly, and the sysupgrade has been used to successfully via the LuCI web interface to update two units to this new build without error.

Per @PolynomialDivision I'd be happy to update to support the "DSA conversion" but it looks like that pull-request is still a Draft? Which seems like it'd further block this long-standing pull-request? :)

sniff122 commented:
Glad to hear you were able to get it to build too

sniff122 commented:
Just done some testing of wireless performance and its significantly worse using openwrt, i get that the drivers wont be the same as what linksys gets for the stock firmware but distance is significantly reduced along with general throughput. On the stock firmware im getting about 500-600mbps (802.11ac ofc) and using openwrt im getting around 200mbps. Both tests were performed in the same room as the AP. I havent really performed any tuning and i honestly dont really know what parameters can be tuned to improve the performance

darkarnium commented:
@sniff122 are you using `ath10k-firmware-qca4019` or `ath10k-firmware-qca4019-ct`?

If using `ath10k-firmware-qca4019` can you try with `ath10k-firmware-qca4019-ct`. If you're using `ath10k-firmware-qca4019-ct` can you recompile with `ath10k-firmware-qca4019` as a test and report back whether you see any improvement between the two?

sniff122 commented:
looks like im using `ath10k-firmware-qca4019-ct` will build with the non `-ct` and report back after testing

sniff122 commented:
looks like im getting about the same throughput with the non `-ct` firmware

sniff122 commented:
Just installed atop with opkg and it looks like the limitation might be coming from IRQ because an IRQ is being hammered when doing a speed test
![img](https://i.sniff122.tech/uploads/c8f4622362b210d42d05.png)

sniff122 commented:
tested with a second AP with the same results

darkarnium commented:
I'm seeing about 281Mbit - 289Mbit average over a 60s `iperf` run with the client on a laptop, and the server ON the ap. On the AP 801.11nac radio has been configured with the SSID the client is connected to, with the client shown as being connected as: `400.0 Mbit/s, 40 MHz, VHT-MCS 9, VHT-NSS 2, Short GI`

<img width="1343" alt="image" src="https://user-images.githubusercontent.com/3048394/169139873-8daa3730-3c5c-4f74-a139-00a8f5934f67.png">

To eliminate bottlenecks due to running `iperf` on the access point I ran the server on a machine with a wired connection to the access point - ableit via a switch. I'm seeing similar `iperf` results in this configuration.

sniff122 commented:
Running iperf between my phone and server on stock firmware:
![img](https://lewisthe.computer/uploads/f47cb411e2328d39da53.png)

running iperf between phone and server on openwrt:
![img](https://lewisthe.computer/uploads/b364eb2d6c8311b43a5e.png)
Actually a little higher than ive had just using the speed test in ubiquiti wifi man (where i pull the 500-600mbps on the stock firmware)

sniff122 commented:
And my phone is connected as: `526.6 Mbit/s, 80 MHz, VHT-MCS 6, VHT-NSS 2` / `780.0 Mbit/s, 80 MHz, VHT-MCS 8, VHT-NSS `2, Short GI with the server on a gigabit ethernet connection to the switch the AP is on

darkarnium commented:
Thanks for those! I'll have a look and see what may be going on. I've upstreamed the board files for this device to the ath10k project, which were pulled directly from the device, but they're also present in this pull-request. I'll need to see what may be different between the stock firmware and the openwrt build. This is quite strange.

sniff122 commented:
alright, hope we can get to the bottom of this. Also ive recently wrote an ifixit teardown guide of the device what could be helpful on the wiki as a guide for users if they want to flash but want images of the process: https://www.ifixit.com/Teardown/Linksys+Velop+WHW01+Teardown/150164

hervadore commented:
Hello, I've had several issues trying to build this for my whw01. Would anyone be able to assist or is there an option to download the file from anywhere?

sniff122 commented:
can you elaborate on the issues you are having? any errors logged? reproduction steps?

hervadore commented:
> can you elaborate on the issues you are having? any errors logged? reproduction steps?

```
git clone https://github.com/openwrt/openwrt.git

cd openwrt

git branch -a

git tag

git checkout v19.07.10

scripts/feeds update -a
scripts/feeds install -a
```

I tried manually making the changes for the patch but assuming I'm doing something wrong there. What would be the best way to apply a patch?

sniff122 commented:
you cant clone the master branch as this PR isnt merged in, here is what i have been doing with success
```
git clone https://github.com/darkarnium/openwrt -b Linksys-WHW01v1
cd openwrt
./scripts/feeds update -a
./scripts/feeds install -a
make menuconfig
make -j $(nproc) kernel_menuconfig
make -j $(nproc) defconfig download clean world
```

for `make menuconfig`, the target system is `Qualcomm Atheros IPQ40xx`, subtarget is `Generic` and target profile is `Linksys WHW01 v1` (feel free to make any other changes you may want)

That should compile successfully assuming you have all needed build dependencies installed, the firmware image you would use to initially flash the WHW01 will be in `bin/targets/ipq40xx/generic/openwrt-ipq40xx-generic-linksys_whw01-v1-squashfs-factory.bin`

hervadore commented:
> you cant clone the master branch as this PR isnt merged in, here is what i have been doing with success
> 
> ```
> git clone https://github.com/darkarnium/openwrt -b Linksys-WHW01v1
> cd openwrt
> ./scripts/feeds update -a
> ./scripts/feeds install -a
> make menuconfig
> make -j $(nproc) kernel_menuconfig
> make -j $(nproc) defconfig download clean world
> ```
> 
> for `make menuconfig`, the target system is `Qualcomm Atheros IPQ40xx`, subtarget is `Generic` and target profile is `Linksys WHW01 v1` (feel free to make any other changes you may want)
> 
> That should compile successfully assuming you have all needed build dependencies installed, the firmware image you would use to initially flash the WHW01 will be in `bin/targets/ipq40xx/generic/openwrt-ipq40xx-generic-linksys_whw01-v1-squashfs-factory.bin`

Thank you so much! That worked. I knew I was doing something wrong.

hervadore commented:
Am I supposed to rename the bin file since the router is looking for nodes-jr.img?

sniff122 commented:
yeah

ddgg commented:
> yeah

thanks for the great work, I want to use it as a router, but there is only lan interface, if I want to add wan, interface should I use eth1? and which of the two eth port should i plug wan cable in?

sniff122 commented:
> > yeah
> 
> thanks for the great work, I want to use it as a router, but there is only lan interface, if I want to add wan, should I use device eth1? and which of the two eth port should i plug wan cable in?

eth0 is the port that connects to the internal switch that handles the 2 ethernet ports. I'm pretty sure you would need to configure a VLAN for WAN in the switch config under network for one of the ports. After you would configure the firewall zone for that VLAN. Someone please correct me if I'm wrong

ddgg commented:
> > > yeah
> > 
> > 
> > thanks for the great work, I want to use it as a router, but there is only lan interface, if I want to add wan, should I use device eth1? and which of the two eth port should i plug wan cable in?
> 
> eth0 is the port that connects to the internal switch that handles the 2 ethernet ports. I'm pretty sure you would need to configure a VLAN for WAN in the switch config under network for one of the ports. After you would configure the firewall zone for that VLAN. Someone please correct me if I'm wrong

yeah, you are right, I add vlan and wan following another router setup and everything is working now, very glad this old but well designed router can be useful again, thank you for the help.

chunkeey commented:
Merged. Thanks!

Sybalau commented:
Hello, can you tell me if this method is compatible with the version socket from WHW01 (WHW01P)?

darkarnium commented:
Hey @Sybalau,

Good question! I've had a quick look around and it looks like the WHW01P uses the same Winbond  W632GU6MB-12 memory, Qualcomm IPQ4018 SoC, CSR8811 Bluetooth radio. _However_, it looks like the 5GHz front-end is different (SKY7482I001 vs SKY85748-11 in the WHW01v1), and the flash also looks like it may be different.

As a result, it seems likely that there would need to be some adjustments to the device-tree for this device, modification of the ethernet switch (as it looks like this device doesn't have any Ethernet ports), and likely extraction of new ath10k calibration data.

All in all, it doesn't seem easily possible - to me at least - to add support for this device to this change set. You may be able to re-use a bunch of the changes in a subsequent pull-request to add support for this device, but there would need to be several other changes based on information extracted from a running device.

georgemoussalem commented:
hello, I've manually applied the patch against 22.03.2 but am struggling to load the board file for wifi to load. From the boot log, I get the following:

[   20.165236] ath10k_ahb a000000.wifi: qca4019 hw1.0 target 0x01000000 chip_id 0x003b00ff sub 0000:0000
[   20.165302] ath10k_ahb a000000.wifi: kconfig debug 0 debugfs 1 tracing 0 dfs 1 testmode 0
[   20.179013] ath10k_ahb a000000.wifi: firmware ver 10.4-3.6-00140 api 5 features no-p2p,mfp,peer-flow-ctrl,btcoex-param,allows-mesh-bcast,no-ps crc32 ba79b746
[   20.199078] ath10k_ahb a000000.wifi: Loading BDF type 0
[   20.199533] ath10k_ahb a000000.wifi: failed to fetch board data for bus=ahb,vendor=0000,device=0000,subsystem-vendor=0000,subsystem-device=0000,variant=linksys-whw01-v from ath10k/QCA4019/hw1.0/board-2.bin
[   20.408068] ath10k_ahb a000000.wifi: failed to fetch board-2.bin or board.bin from ath10k/QCA4019/hw1.0
[   20.408157] ath10k_ahb a000000.wifi: failed to fetch board file: -12
[   20.416603] ath10k_ahb a000000.wifi: could not probe fw (-12)
[   21.338513] ath10k_ahb a800000.wifi: qca4019 hw1.0 target 0x01000000 chip_id 0x003b00ff sub 0000:0000
[   21.338584] ath10k_ahb a800000.wifi: kconfig debug 0 debugfs 1 tracing 0 dfs 1 testmode 0
[   21.350744] ath10k_ahb a800000.wifi: firmware ver 10.4-3.6-00140 api 5 features no-p2p,mfp,peer-flow-ctrl,btcoex-param,allows-mesh-bcast,no-ps crc32 ba79b746
[   21.372371] ath10k_ahb a800000.wifi: Loading BDF type 0
[   21.372624] ath10k_ahb a800000.wifi: failed to fetch board data for bus=ahb,vendor=0000,device=0000,subsystem-vendor=0000,subsystem-device=0000,variant=linksys-whw01-v from ath10k/QCA4019/hw1.0/board-2.bin
[   21.597987] ath10k_ahb a800000.wifi: failed to fetch board-2.bin or board.bin from ath10k/QCA4019/hw1.0
[   21.598059] ath10k_ahb a800000.wifi: failed to fetch board file: -12
[   21.606609] ath10k_ahb a800000.wifi: could not probe fw (-12)

any advise would be highly appreciated!

Nyshan commented:
I'm currently trying to get DSA support setup for this device but I'm not having much luck. Could anyone assist me?
